hpl-2.2/man/man3/HPL_dswap.3
.TH HPL_dswap 3 "February 24, 2016" "HPL 2.2" "HPL Library Functions"
.SH NAME
HPL_dswap \- y <-> x.
.SH SYNOPSIS
\fB\&#include "hpl.h"\fR
 
\fB\&void\fR
\fB\&HPL_dswap(\fR
\fB\&const int\fR
\fI\&N\fR,
\fB\&double *\fR
\fI\&X\fR,
\fB\&const int\fR
\fI\&INCX\fR,
\fB\&double *\fR
\fI\&Y\fR,
\fB\&const int\fR
\fI\&INCY\fR
\fB\&);\fR
.SH DESCRIPTION
\fB\&HPL_dswap\fR
swaps the vectors x and y.
.SH ARGUMENTS
.TP 8
N       (local input)           const int
On entry, N specifies the length of the vectors  x  and  y. N
must be at least zero.
.TP 8
X       (local input/output)    double *
On entry,  X  is an incremented array of dimension  at  least
( 1 + ( n - 1 ) * abs( INCX ) )  that  contains the vector x.
On exit, the entries of the incremented array  X  are updated
with the entries of the incremented array Y.
.TP 8
INCX    (local input)           const int
On entry, INCX specifies the increment for the elements of X.
INCX must not be zero.
.TP 8
Y       (local input/output)    double *
On entry,  Y  is an incremented array of dimension  at  least
( 1 + ( n - 1 ) * abs( INCY ) )  that  contains the vector y.
On exit, the entries of the incremented array  Y  are updated
with the entries of the incremented array X.
.TP 8
INCY    (local input)           const int
On entry, INCY specifies the increment for the elements of Y.
INCY must not be zero.
.SH EXAMPLE
\fI\&#include "hpl.h"\fR
 
int main(int argc, char *argv[])
.br
{
.br
   double x[3], y[3];
.br
   x[0] = 1.0; x[1] = 2.0; x[2] = 3.0;
.br
   y[0] = 4.0; y[1] = 5.0; y[2] = 6.0;
.br
   HPL_dswap( 3, x, 1, y, 1 );
.br
   printf("x=[%f,%f,%f]\en", x[0], x[1], x[2]);
.br
   printf("y=[%f,%f,%f]\en", y[0], y[1], y[2]);
.br
   exit(0); return(0);
.br
}
.SH SEE ALSO
.BR HPL_daxpy \ (3),
.BR HPL_dcopy \ (3),
.BR HPL_dscal \ (3).


hpl-2.2/HISTORY
==============================================================
 High Performance Computing Linpack Benchmark (HPL)
 HPL - 2.2 - February 24, 2016
==============================================================

 History

 - 09/09/00 Public release of Version 1.0

 - 09/27/00 A couple of mistakes in the  VSIPL  port have been
 corrected.  The tar file as well as the web site were updated
 on September 27th, 2000.  Note  that  these problems were not
 affecting the BLAS version of the software in any way.

 - 01/01/04 Version 1.0a
 The  MPI  process grid  numbering  scheme  is now an run-time
 option.
 The inlined assembly  timer  routine that caused the compila-
 tion to fail when using  gcc  version 3.3  and above has been
 removed from the package.
 Various building problems on the T3E have been fixed;  Thanks
 to Edward Anderson.

 - 15/12/04 Version 1.0b
 Weakness of the pseudo-random matrix generator found for pro-
 blem sizes being power of twos and larger  than 2^15;  Thanks
 to Gregory Bauer. This problem has not been fixed. It is thus
 currently recommended to  HPL  users willing to test matrices
 of size larger than 2^15 to not use power twos.

 When the matrix size is such that one needs  > 16 GB  per MPI
 rank,  the  intermediate  calculation  (mat.ld+1) * mat.nq in
 HPL_pdtest.c  ends up  overflowing  because  it is done using
 32-bit arithmetic.  This issue has been fixed by  typecasting
 to size_t; Thanks to John Baron.

 - 09/10/08 Version 2.0

 Piotr Luszczek changed to 64-bit RNG, modified files:
 -- [M] include/hpl_matgen.h
 -- [M] testing/matgen/HPL_ladd.c
 -- [M] testing/matgen/HPL_lmul.c
 -- [M] testing/matgen/HPL_rand.c
 -- [M] testing/ptest/HPL_pdinfo.c

 For a motivation for the change, see:
    Dongarra and Langou, ``The Problem with the Linpack
    Benchmark Matrix Generator'', LAWN 206, June 2008.

 -- [M] testing/ptest/HPL_pdtest.c  --

 Julien Langou changed the test for correctness from
      ||Ax-b||_oo / ( eps * ||A||_1  * N            )
      ||Ax-b||_oo / ( eps * ||A||_1  * ||x||_1      )
      ||Ax-b||_oo / ( eps * ||A||_oo * ||x||_oo * N )
 to the normwise backward error
      || r ||_oo / ( eps * ( || x ||_oo * || A ||_oo + || b ||_oo ) * N )
 See:
  Nicholas J. Higham, ``Accuracy and Stability of Numerical Algorithms'',
  Society for Industrial and Applied Mathematics, Philadelphia, PA, USA,
  Second Edition, pages = xxx+680, ISBN = 0-89871-521-0, 2002.

 Note that in our case || b ||_oo is almost for sure
 1/2, we compute it anyway.

 - 10/26/2012 Version 2.1

 Piotr Luszczek introduced exact time stamping for HPL_pdgesv():
 -- [M] dist/include/hpl_misc.h
 -- [M] dist/testing/ptest/HPL_pdtest.c

 Piotr Luszczek fixed out-of-bounds access in data spreading functions
 and exact time stamping for HPL_pdgesv():
 -- [M] dist/src/pgesv/HPL_spreadN.c
 -- [M] dist/src/pgesv/HPL_spreadT.c
 Thanks to Stephen Whalen from Cray.

 - 02/24/2016 Version 2.2

 Piotr Luszczek added continuous reporting of factorization progress
 submitted by Intel and make scripts that uses Intel software tools and
 libraries and their Apple's Mac OS X equivalents.


hpl-2.2/src/blas/HPL_dswap.c
/* 
 * -- High Performance Computing Linpack Benchmark (HPL)                
 *    HPL - 2.2 - February 24, 2016                          
 *    Antoine P. Petitet                                                
 *    University of Tennessee, Knoxville                                
 *    Innovative Computing Laboratory                                 
 *    (C) Copyright 2000-2008 All Rights Reserved                       
 *                                                                      
 * -- Copyright notice and Licensing terms:                             
 *                                                                      
 * Redistribution  and  use in  source and binary forms, with or without
 * modification, are  permitted provided  that the following  conditions
 * are met:                                                             
 *                                                                      
 * 1. Redistributions  of  source  code  must retain the above copyright
 * notice, this list of conditions and the following disclaimer.        
 *                                                                      
 * 2. Redistributions in binary form must reproduce  the above copyright
 * notice, this list of conditions,  and the following disclaimer in the
 * documentation and/or other materials provided with the distribution. 
 *                                                                      
 * 3. All  advertising  materials  mentioning  features  or  use of this
 * software must display the following acknowledgement:                 
 * This  product  includes  software  developed  at  the  University  of
 * Tennessee, Knoxville, Innovative Computing Laboratory.             
 *                                                                      
 * 4. The name of the  University,  the name of the  Laboratory,  or the
 * names  of  its  contributors  may  not  be used to endorse or promote
 * products  derived   from   this  software  without  specific  written
 * permission.                                                          
 *                                                                      
 * -- Disclaimer:                                                       
 *                                                                      
 * THIS  SOFTWARE  IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,  INCLUDING,  BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE UNIVERSITY
 * OR  CONTRIBUTORS  BE  LIABLE FOR ANY  DIRECT,  INDIRECT,  INCIDENTAL,
 * SPECIAL,  EXEMPLARY,  OR  CONSEQUENTIAL DAMAGES  (INCLUDING,  BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA OR PROFITS; OR BUSINESS INTERRUPTION)  HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT,  STRICT LIABILITY,  OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 * ---------------------------------------------------------------------
 */ 
/*
 * Include files
 */
#include "hpl.h"

#ifndef HPL_dswap

#ifdef STDC_HEADERS
void HPL_dswap
(
   const int                        N,
   double *                         X,
   const int                        INCX,
   double *                         Y,
   const int                        INCY
)
#else
void HPL_dswap
( N, X, INCX, Y, INCY )
   const int                        N;
   double *                         X;
   const int                        INCX;
   double *                         Y;
   const int                        INCY;
#endif
{
/* 
 * Purpose
 * =======
 *
 * HPL_dswap swaps the vectors x and y.
 * 
 *
 * Arguments
 * =========
 *
 * N       (local input)                 const int
 *         On entry, N specifies the length of the vectors  x  and  y. N
 *         must be at least zero.
 *
 * X       (local input/output)          double *
 *         On entry,  X  is an incremented array of dimension  at  least
 *         ( 1 + ( n - 1 ) * abs( INCX ) )  that  contains the vector x.
 *         On exit, the entries of the incremented array  X  are updated
 *         with the entries of the incremented array Y.
 *
 * INCX    (local input)                 const int
 *         On entry, INCX specifies the increment for the elements of X.
 *         INCX must not be zero.
 *
 * Y       (local input/output)          double *
 *         On entry,  Y  is an incremented array of dimension  at  least
 *         ( 1 + ( n - 1 ) * abs( INCY ) )  that  contains the vector y.
 *         On exit, the entries of the incremented array  Y  are updated
 *         with the entries of the incremented array X.
 *
 * INCY    (local input)                 const int
 *         On entry, INCY specifies the increment for the elements of Y.
 *         INCY must not be zero.
 *
 * ---------------------------------------------------------------------
 */ 
#ifdef HPL_CALL_CBLAS
   cblas_dswap( N, X, INCX, Y, INCY );
#endif
#ifdef HPL_CALL_VSIPL
   register double           x0, x1, x2, x3, y0, y1, y2, y3;
   double                    * StX;
   register int              i;
   int                       nu;
   const int                 incX2 = 2 * INCX, incY2 = 2 * INCY,
                             incX3 = 3 * INCX, incY3 = 3 * INCY,
                             incX4 = 4 * INCX, incY4 = 4 * INCY;

   if( N > 0 )
   {
      if( ( nu = ( N >> 2 ) << 2 ) != 0 )
      {
         StX = X + nu * INCX;
 
         do
         {
            x0 = (*X);      y0 = (*Y);      x1 = X[INCX ];  y1 = Y[INCY ];
            x2 = X[incX2];  y2 = Y[incY2];  x3 = X[incX3];  y3 = Y[incY3];
            *Y        = x0; *X        = y0; Y[INCY ]  = x1; X[INCX ]  = y1;
            Y[incY2]  = x2; X[incX2]  = y2; Y[incY3]  = x3; X[incX3]  = y3;
            X += incX4; Y += incY4;
 
         } while( X != StX );
      }
 
      for( i = N - nu; i != 0; i-- )
      { x0  = (*X); y0  = (*Y); *Y = x0; *X = y0; X += INCX; Y += INCY; }
   }
#endif
#ifdef HPL_CALL_FBLAS
#ifdef HPL_USE_F77_INTEGER_DEF
   const F77_INTEGER         F77N = N, F77incx = INCX, F77incy = INCY;
#else
#define F77N                 N
#define F77incx              INCX
#define F77incy              INCY
#endif
   F77dswap( &F77N, X, &F77incx, Y, &F77incy );
#endif
/*
 * End of HPL_dswap
 */
}
 
#endif
